FOCUS #231: ListUnitPrice metric review (#234)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: supporting_content/metrics/listunitprice.md

### GCP

* [Structure of pricing data export | Cloud Billing](https://cloud.google.com/billing/docs/how-to/export-data-bigquery-tables/pricing-data)
* [Get Google Cloud pricing information | Cloud Billing (GCP Cloud Billing API)](https://cloud.google.com/billing/docs/how-to/get-pricing-information-api)
* [Structure of Detailed data export | Cloud Billing](https://cloud.google.com/billing/docs/how-to/export-data-bigquery-tables/detailed-usage)

### Microsoft

* [Understand usage details fields - Microsoft Cost Management](https://learn.microsoft.com/en-us/azure/cost-management-billing/automate/understand-usage-details-fields)
* [Azure Retail Prices REST API overview | Microsoft Learn](https://learn.microsoft.com/en-us/rest/api/cost-management/retail-prices/azure-retail-prices)

### OCI

* [Cost and Usage Reports Overview](https://docs.oracle.com/en-us/iaas/Content/Billing/Concepts/usagereportsoverview.htm)
* [Estimate Your Monthly Cost (List Pricing REST API)](https://docs.oracle.com/en-us/iaas/Content/GSG/Tasks/signingup_topic-Estimating_Costs.htm#accessing_list_pricing)

## Example usage scenarios

> * See [Appendix: Quantity-Driven Pricing strategies](../apendix/quantity_driven_pricing_strategies.md) section for more information on

Reviewer: that should be "./appendix/quantity_driven_pricing_strategies.md" (note "appendix")

  * Quantity-Driven Pricing strategies (such as flat-rate, volume-based rates and tier-based rating)
  * related UC scenarios
  * Current values observed in billing data for various scenarios
  * Alternative data sources for various scenarios
* See [Pricing Support – UCs and Data samples Spreadsheet](https://docs.google.com/spreadsheets/d/1AZ-vtkKeKwYc8rqhxP1zMTnAVAS-svmWQQmr8cpv-IM/edit#gid=117987709) for additional UC scenarios and sample data.

## Discussion / Scratch space

### Prices and Currencies

> * We deliberated on whether to specify both BillingCurrency and PricingCurrency and decided to enforce a single currency, specifically BillingCurrency. This approach provides consistency and simplifies invoice reconciliation. Since some providers don't include List Unit Prices in BillingCurrency in their public price sheets, the inclusion of a CurrencyExchangeRate dimension in the billing data becomes imperative (name subject to change). This is necessary to ensure we can accurately compare and match ListUnitPrices provided in billing data with those published in public price sheets.

Reviewer 2: This kind of hides the major point here... FOCUS data will NOT be consistent with pricing APIs or invoices, which show prices in the pricing currency.

## Discussion / Scratch space

### Prices and Currencies

* We deliberated on whether to specify both BillingCurrency and PricingCurrency and decided to enforce a single currency, specifically BillingCurrency. This approach provides consistency and simplifies invoice reconciliation. Since some providers don't include List Unit Prices in BillingCurrency in their public price sheets, the inclusion of a CurrencyExchangeRate dimension in the billing data becomes imperative (name subject to change). This is necessary to ensure we can accurately compare and match ListUnitPrices provided in billing data with those published in public price sheets.

### Free-tier, volume/tier-based, BYOL-based and dynamically-priced SKU rates

* Being inclusive of free-tier, volume/tier-based, BYOL-based and dynamically priced SKU rates, the List Unit Price cannot be used for calculating savings based on volume/tier-based pricing, the use of pre-owned software licenses (BYOL - Bring Your Own License), leveraging interruptible resources and/or services, or optimizing usage to take advantage of dynamic pricing models.

### Current data sources

> * For GCP and OCI, Pricing Data serves as the sole data source for ListUnitPrice, while for AWS, it is the preferred data source due to concerns related to volume/tier-based pricing. To determine the relevant price in Pricing Data for a specific charge record, the billing data must include the UnitPriceId. As previously mentioned, when BillingCurrency and PricingCurrency differ, the exchange rate must also be considered.

Reviewer 3: > the billing data must include the UnitPriceId.

I understood the final name for UnitPriceId is actually SKUPriceId.

Author: Thank you for pointing this out! The note refers to a discussion that took place long before the SkuPriceId column was named and specified. An update is coming soon.

Author: supporting content updated (a bit more details provided).
resolving the conversation

> * For GCP and OCI, Pricing Data serves as the sole data source for ListUnitPrice, while for AWS, it is the preferred data source due to concerns related to volume/tier-based pricing. To determine the relevant price in Pricing Data for a specific charge record, the billing data must include the UnitPriceId. As previously mentioned, when BillingCurrency and PricingCurrency differ, the exchange rate must also be considered.

Author: ```suggestion
* For GCP and OCI, Pricing Data serves as the sole data source for ListUnitPrice, while for AWS, it is the preferred data source due to concerns related to volume/tier-based pricing. To determine the relevant price in Pricing Data for a specific charge record, the billing data must include the SkuPriceId. As previously mentioned, when BillingCurrency and PricingCurrency differ, the exchange rate must also be considered.
```

| :-------- | :------------------------------------------------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | AWS       | CUR;<br>AWS Price List API                         | the closest match - pricing/publicOnDemandRate: The public On-Demand Instance rate in this billing period for the specific row of usage. If you have SKUs with multiple On-Demand public rates, the equivalent rate for the highest tier is displayed. For example, services offering free-tiers or tiered pricing.<br>_Note: Since currently only pricing/publicOnDemandRate definition includes note regarding highest tier, I'm not sure if volume-based pricing applies to all unit prices (rates) in CUR files. If it's not consistent, we might have to rely on the AWS Price List API._ |

Reviewer 4: I'm not sure how formal/informal we can be in these supporting content docs -- but the use of first and second person here (i.e. "I" and "you") feels off to me.  I would suggest using whatever convention we have for this -- and, if we lack any such convention, sidestepping any use of first and second person with some combination of nouns and passive voice would be preferable to me.

```suggestion
| AWS       | CUR;<br>AWS Price List API                         | the closest match - pricing/publicOnDemandRate: The public On-Demand Instance rate in this billing period for the specific row of usage. If there are SKUs with multiple On-Demand public rates, the equivalent rate for the highest tier is displayed. For example, services offering free-tiers or tiered pricing.<br>_Note: Since currently only pricing/publicOnDemandRate definition includes note regarding highest tier, it is unclear whether volume-based pricing applies to all unit prices (rates) in CUR files. If it's not consistent, the practitioner might have to rely on the AWS Price List API._ |
```
